docs/sql-migration-guide.md

```diff
@@ -185,6 +185,8 @@ license: |
     * `ALTER TABLE .. ADD PARTITION` throws `PartitionsAlreadyExistException` if new partition exists already
     * `ALTER TABLE .. DROP PARTITION` throws `NoSuchPartitionsException` for not existing partitions
 
+  - In Spark 3.1, when bucket join is enabled(`spark.sql.sources.bucketing.enabled=true`), whether to do bucketed scan on input tables is decided automatically based on query plan. Bucketed scan is not used if 1. query does not have operators to utilize bucketing (e.g. join, group-by) or 2. there's an exchange between these operators and table scan. You can restore old behavior by setting `spark.sql.sources.bucketing.autoBucketedScan.enabled` to `false`.  
```

Comment:
Hm, this was technically a performance improvement which shouldn't cause user-facing behaviour impacts. I think we should focus on user-facing changes here. Does it cause any user-facing change?

Reply:
Yes. In my case table `t` is bucketed on `i` and sorted on `j`.  The following SQL takes 6 mins to run in Spark 3.1.1 while 30s in Spark 2.3.1.

```sql
select k
from t
where j='a'
limit 10
```

From the log I could see it's bucketed scan in 2.3.1 while non-bucketed scan 3.1.1 but I didn't know why. It took me days to debug till I found this config. If it were documented in migration guide, I could have figured it out much sooner. Hence, I think it might help others as well.

Reply:
This is a performance improvement, which should be listed in the release notes, not the migration guide.

Migration guide is used for changes that can break user queries, or make them return a different result.

Reply:
I think this is a performance regression we should fix. Seems it's better to use bucketed scan if we have predicate against bucket or sort columns. cc @c21

Reply:
Sure, I think it's doable. Let me change the rule itself to be more restricted

Reply:
btw @manuzhang - just to double check, if the query has `LIMIT`, then the rule will not disable bucketed scan, as limit operator is not allowed operator to disable bucketed scan ([code](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/bucketing/DisableUnnecessaryBucketedScan.scala#L134)). Could you come up with a complete query example for us to check?

Reply:
This is the complete query. I think it hits the `FileSourceScanExec` case https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/bucketing/DisableUnnecessaryBucketedScan.scala#L100

Reply:
@manuzhang - please see my comment above - https://github.com/apache/spark/pull/35514#discussion_r807540227 . Thanks.

Reply:
`sql("SELECT * FROM t1 WHERE k = 1 LIMIT 10")` should trigger the issue?

Reply:
Even if the predicate is not bucket or sort column, this query could still be much slower without bucketed scan for large bucket file (e.g. 3GB) since more tasks and stages will be launched instead of 1.

Reply:
@c21 Add bucket column to project or just select all as @cloud-fan suggested. 

```
select i, k
from t
where j='a'
limit 10
```

Reply:
> Even if the predicate is not bucket or sort column, this query could still be much slower without bucketed scan for large bucket file (e.g. 3GB) since more tasks and stages will be launched instead of 1.

Why bucket scan can help here? Without bucket scan, Spark can smartly decide the parallelism w.r.t. the data size, while with bucket scan, the number of partitions must be the same as number of buckets, which is usually not optimal.

Reply:
I mean in the case of `limit` where Spark incrementally launches tasks in stages to find rows satisfying the predicate.

Reply:
If selecting bucket columns, I concur this is an issue caused by the rule. I am working on a fix to not disable bucketed scan if having filter on sorted columns. I guess the trick of bucket scan to be more efficient for limit query is:
* with bucketed scan: one bucket file is processed by one task. For large bucket file, small limit value, and filter on sorted column, it can return early with skipping scan a lot of rest of rows in the file.
* without bucketed scan: one bucket file can be split to multiple tasks if it's large. The filter for some of file splits may need to scan all of rows in the split without skipping, because the sorted file is split into multiple splits.

Reply:
As commented above, filter on non-sorted column can also be much faster with bucketed scan. Disabling bucketed scan having filter on sorted columns won't fix all regressions in the case of `limit`.

Reply:
> filter on non-sorted column can also be much faster with bucketed scan

I don't get this argument, do you mind giving a concrete example with metrics?

Reply:
It's not easy to come up with a simple example. Here are some stats of our production case.
Table `t` has 10000 buckets, bucketed on column i and sorted on column j,  and each bucket file is around 3GB.
There are more than 10 rows satisfying `x='a'` at the last part in `part-00000-*.parquet` for below query.

```sql
select i, k
from t
where x = 'a'
limit 10;
```

With bucketed scan, all results can be found in `part-00000-*.parquet` in one task.
Without bucketed scan (`spark.sql.files.maxPartitionBytes=128MB`),  1 task, 4 tasks, 20 tasks,..., 151875 tasks were launched in 9 stages to find all results.
IIUC, file splits are sorted by length from large to small. Hence the last part of `part-00000-*.parquet` was only scanned in last stage.

Reply:
> With bucketed scan, all results can be found in part-00000-*.parquet in one task.

This seems more like a coincidence. It's nothing about your bucket/sort column. Maybe bucketed scan is good for this particular table this time, but maybe not next time when the data of this table has been changed.

Reply:
I'm not arguing about bucket scan is always faster but disabling bucket scan automatically should not introduce unexpected regression or have documentation about it.

Reply:
@manuzhang - regarding documentation, it's already added in https://spark.apache.org/docs/latest/configuration.html .

<img width="1630" alt="Screen Shot 2022-02-16 at 1 01 28 PM" src="https://user-images.githubusercontent.com/4629931/154356460-1b97d56a-9500-4b0d-9d26-b59da7247cfe.png">

Reply:
@c21 Thanks. That's good when the change is transparent to users. In this case, however, it's impacting users when upgrading to Spark 3.1. And the first place I'd check is migration guide.

Reply:
Again, this only affects perf, so we don't put it in the migration guide. The impact is low because it only affects tables with special data layouts.

I'd recommend including `x` as bucket columns as well, as you assume `x = 'a'` hits only one file.

Reply:
Okay. In our case, the bucket table is used by many users so it's not plausible to change bucket column for one use case. I think all perf regressions regardless of data layouts need be fixed for this optimization.